Acceptance process delegation for most items + exceptions (#131)
cncf/cnf-wg: CONTRIBUTING.md

Reviews and comments on open [Pull Requests](https://github.com/cncf/cnf-wg/pulls) are also appreciated.

#### Steps for creating a new PR:

- Once you have made your change or added new content
- Ensure you are up to date with the `main` branch of the cnf-wg repository
- Open a new [Pull Request](https://github.com/cncf/cnf-wg/pulls)

> - Choose at least 5 reviewers OR choose a [co-chair](https://github.com/cncf/cnf-wg/blob/main/GOVERNANCE.md#chairs) who will find reviewers for you

Reviewer: suggestion to change link from (https://github.com/cncf/cnf-wg/blob/main/GOVERNANCE.md#chairs) to (GOVERNANCE.md#chairs)

> - Choose at least 5 reviewers OR choose a [co-chair](https://github.com/cncf/cnf-wg/blob/main/GOVERNANCE.md#chairs) who will find reviewers for you

Reviewer: ```suggestion
- Choose at least 5 reviewers OR choose a [co-chair](GOVERNANCE.md#chairs) who will find reviewers for you
```

Reviews and comments on open [Pull Requests](https://github.com/cncf/cnf-wg/pulls) are also appreciated.

> #### Acceptance of PRs
> 
> - Spelling, grammar, and new interested parties require 1 reviewer
> - Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may accepted after approval of 5 community reviewers.

Reviewer 2: ```suggestion
#### Acceptance of PRs for merging

- Spelling, grammar, and new interested parties require 1 apporval for merging
- Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may merged after approval of 5 community reviewers.
```

#### Steps for creating a new PR

- Once you have made your change or added new content
- Ensure you are up to date with the `main` branch of the cnf-wg repository
- Open a new [Pull Request](https://github.com/cncf/cnf-wg/pulls)
- Choose at least 5 reviewers OR choose a [co-chair](GOVERNANCE.md#chairs) who will find reviewers for you

#### Steps to accept a PR

1. Reviewers will review the PR and give their feedback: approve, request change, comment w/o approval
1. After required number of approvals from reviewers is reached the PR may be merged

Anyone with merge access can merge the PR after the item has been approved.

### Acceptance criteria for Contributions

Reviewers - Any community member

> - Everyone in the community is able to and encouraged to do reviews

Reviewer 2: ```suggestion
- Everyone in the community is able to and encouraged to do reviews. Anyone can be a part of the CNF WG community.
```

Minor changes:

- Spelling, grammar, and adding new interested parties require 1 reviewer

Majority of other changes:

> - Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may be accepted after approval of 5 community reviewers.

Reviewer 2: ```suggestion
- Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may be merged after approval of 5 community reviewers.
```

#### Steps for creating a new PR:

- Once you have made your change or added new content
- Ensure you are up to date with the `main` branch of the cnf-wg repository
- Open a new [Pull Request](https://github.com/cncf/cnf-wg/pulls)
- Choose at least 5 reviewers OR choose a [co-chair](GOVERNANCE.md#chairs) who will find reviewers for you

#### Steps to accept a PR

1. Reviewers will review the PR and give their feedback: approve, request change, comment w/o approval
1. After required number of approvals from reviewers is reached the PR may be merged

Anyone with merge access can merge the PR after the item has been approved.

### Acceptance criteria for Contributions

Reviewers - Any community member
- Everyone in the community is able to and encouraged to do reviews

Minor changes:
- Spelling, grammar, and adding new interested parties require 1 reviewer

Majority of other changes:
- Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may be accepted after approval of 5 community reviewers.

> <<<<<<< HEAD

Reviewer 3: oopsy

Author: oops

Reviews and comments on open [Pull Requests](https://github.com/cncf/cnf-wg/pulls) are also appreciated. 

##### Steps for creating a new PR:
- Once you have made your change or added new content
- Ensure you are up to date with the `main` branch of the cnf-wg repository
- Open a new [Pull Request](https://github.com/cncf/cnf-wg/pulls)

> - Tag at least 5 reviewers

Reviewer 2: Where do people find those reviewers? If someone new is coming to the repo, they won't know who to tag

Reviewer 4: Could we add an LFR (Looking For Reviewers) label?

Author: @xmulligan ideally we could use github teams to add all reviewers to the team. Then you can tag the whole group like: @cncf/cnf-wg-reviewers. Unfortunately teams only allow org members and not "outside collaborators".  

Someone new can click the review button on the PR and select folks, but we do have people listed who may not be active or just happen to have access via the org level permissions.  
![image](https://user-images.githubusercontent.com/26697/114221046-63fef580-9932-11eb-9237-6b3d39d4c539.png)

One option is to create a new org for all of the CNF initiatives. We thought about this in the past, but only had cnf-testbed at the time. Now there are 4.  If we did this we could create teams and add everyone who is participating to the new "CNCF CNF" org.

An alternative approach to deal with the github limitations are to to create a file in the repo we can add the github user handle (and probably name) for anyone that would like to be a reviewer. K8s and many projects uses a file called OWNERS to list reviewers and other roles for tagging.

From there someone can manually look at the file and pick some user names to tag (with the reviewers button or via comment) or we can add a bot like many projects have which supports using an owners file.

Many projects and communities use automation with something like an OWNERS file including the K8s community.  They use prow and the bot has a ton of features -- https://github.com/kubernetes/test-infra/blob/master/prow/README.md.  We can probably go with something simpler if we want to automate things.

Github supports some automation via a CODEOWNERS file, but it has limitations such as automatically notifying everyone that is listed.  That could get spammy of course if everyone is notified for every new PR to review vs a subset. See https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners

**TL;DR**
- Couple of options
    - Create a file with a list of the reviewers
    - Create new CNCF CNF org (eg. cncf-cnf) and use org teams
- Decide on manual or automation

Author: @jeffsaelens yes for a label.  Who will look for the label though?  If we automate or its manually noticed and an action to choose reviewers happens when the label is there we still need a list of reviewers somewhere.

Reviewer 3: To get us started (until we find a nice technical solution or something else) I would basically say 'tag at least 5 reviewers OR tag a co-chair who will find reviewers for you'.  It doesn't have to be perfect just yet, only functional.

Reviewer 3: Indeed: if we make a team of 'standard reviewers' it can be their first responsibility to spread the joy to others (which works slightly better than a label would, imo)

Reviewer 5: Is this step a must?

As GH sends notification emails, I wonder if the notification about tagging will be lost or rather just a duplicate of the notification about the new PR. Would this be better as a step to take if the PR sits unreviewed for longer than X?

Author: @ildikov do you mean comment tagging (team or individuals) or something else

I was taking @iawells comment
>tag at least 5 reviewers OR tag a co-chair who will find reviewers for you'
to mean typing the usernames into the choose reviewers box on the top right of a pull request re: screenshot https://github.com/cncf/cnf-wg/pull/131#discussion_r610820564 with _"Request up to 15 reviewers"_

Author: @iawells @ildikov see latest update to CONTRIBUTING guide:
- Choose at least 5 reviewers OR choose a co-chair who will find reviewers for you

#### Steps to accept a PR

> 1. Reviewers will review the PR and give their feedback: approve, request change, comment w/o approval

Reviewer 2: Who are "Reviewers"? Who can be one?

Author: @xmulligan any community member can be a reviewer.

Maybe it should say Community Reviewers or Community Member?

We can grant github read access to the repo to allow the reviewer to be tagged in a PR in the reviews section.

Reviewer 3: I see potential reviewers as any community member for the moment.  We can get more specific in the future, but for the moment this lets us get moving (and is, in practice, writing down what we're already doing).

Reviewer 5: If we have a list of identified roles I think it's better to document that somewhere and refer to that list where needed as opposed to trying to create descriptive terms like 'community reviewer'. As everyone in the community should be able and encouraged to review, saying 'community reviewer' creates a confusing distinction as I don't think we will have non-community reviewers.

Author: @iawells @ildikov see latest update to contributing guide:

Reviewers - Any community member
- Everyone in the community is able to and encouraged to do reviews

1. After required number of approvals from reviewers is reached the PR may be merged

Anyone with merge access can merge the PR after the item has been approved.

### Acceptance of Contributions

Minor changes:
- Spelling, grammar, and adding new interested parties require 1 reviewer

Majority of other changes:

> - Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may accepted after approval of 5 community reviewers.

Reviewer 2: Who a "community reviewer" is needs to be defined somewhere. What about adding a requirement that they come from at least 3 different companies?

Author: Yes for a definition being somewhere.  I was thinking anyone in the community can say they would like to review.  So this is not a select set for the default reviews.  Open to anyone in the community. We can add other types of reviewers later if needed.

Not sure that its necessary at this time to limit reviews from the same org/company yet for the items not in the Additional Approval Items list.  We could modify the CONTRIBUTING guide when we see a problem or after further thought. This path was to keep it simple for the defaults.

Specifically for non-controversial items we want to lower the barrier for making contributions.  If the barrier is too low we adjust later. If something needs more discussion, eg. its something that is more "controversial", then maybe it gets added to the other list.

Reviewer 4: I think the last paragraph is what's important. For basic house keeping, we should make things simple to keep things moving smoothly. Company separation for grammar/spelling fixes, or adding events to a list somewhere etc... don't need any special protection. 

As we begin to produce best practices, and things that have impact, we can tighten up what a role to approve merges would look like.

Author: @xmulligan see [latest update to contributing guide](https://github.com/cncf/cnf-wg/blob/minimal-approval-process-with-exception-list/CONTRIBUTING.md#acceptance-of-contributions)

Reviewers - Any community member
- Everyone in the community is able to and encouraged to do reviews

> - Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may accepted after approval of 5 community reviewers.

Reviewer 5: ```suggestion
- Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may be accepted after approval of 5 community reviewers.
```

Reviews and comments on open [Pull Requests](https://github.com/cncf/cnf-wg/pulls) are also appreciated. 

##### Steps for creating a new PR:
- Once you have made your change or added new content
- Ensure you are up to date with the `main` branch of the cnf-wg repository
- Open a new [Pull Request](https://github.com/cncf/cnf-wg/pulls)

> - Choose at least 5 reviewers OR choose a co-chair who will find reviewers for you

Reviewer 6: Given the current co-chairs are listed in the governance document maybe they can be referenced
```suggestion
- Choose at least 5 reviewers OR choose a [co-chair](https://github.com/cncf/cnf-wg/blob/main/GOVERNANCE.md#chairs) who will find reviewers for you
```

#### Steps to accept a PR
1. Reviewers will review the PR and give their feedback: approve, request change, comment w/o approval

> 1. After required number of approvals from reviewers is reached the PR may be merged

Reviewer 7: Should there be an add that the issues/questions raised have been addressed? Co-chair should ensure prior to merge -- co-chairs responsible for merge or insist a co-chair approval.

Reviewer 7: Are approvals to be ignored that were made prior to non-trivial changes?

Author: @pgoyal01 I'm trying to go through and address all comments in the PR.

If you want it in the CONTRIBUTING.md then please open a new issue and/or PR.

Also remember we can always go back and open a new PR to address anything we consider a mistake or now want to see differently.

Anyone with merge access can merge the PR after the item has been approved.

> ### Acceptance of Contributions

Reviewer 6: ```suggestion
### Acceptance criteria for Contributions
```

Reviewers - Any community member
- Everyone in the community is able to and encouraged to do reviews

Minor changes:

> - Spelling, grammar, and adding new interested parties require 1 reviewer

Reviewer 6: Maybe we can apply [lazy consensus](https://community.apache.org/committers/lazyConsensus.html) here

Reviewer 7: Should there be a restriction that the approver is not from same company as the author

Author: @pgoyal01 that might be a good idea. @xmulligan had a related suggestion about reviewers.  

For now I suggest we leave it more open, then modify if we see it as a problem.  The CONTRIBUTING.md can be updated based on the criteria we list here.  Eg. 5 community reviewers.

Majority of other changes:

> - Other changes with exception to those listed in the [Governance Additional Approval Items list](GOVERNANCE.md#additional-approval-items) may be accepted after approval of 5 community reviewers.

Reviewer 7: Should there be a requirement on the diversity of reviewers (such as not more than 2 from the same company, at least X from each company industry, not from the PR author company, etc,

Author: @pgoyal01 this could be brought up as a new issue, discussion, or PR.
